### PR TITLE
Entities and repos for Juror Summons Replies

### PR DESCRIPTION
**Pull request by Matt-Smith-CGI**
**Before creating a pull request make sure that:**

- [X] commit messages are meaningful and follow good commit message guidelines

### Change description ###

-Added relevant classes for Juror Summons Reply
-Added generator tags to relevant fields
-Added relevant repos for Juror Summons

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import java.io.Serializable;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+
+
+@MappedSuperclass
+@Table(name = "juror_response", schema = "juror_mod")
+@Getter
+@Setter
+
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: Space not needed

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+public class AbstractJurorResponse extends Address implements Serializable {
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: Should be an abstract class as we do not want this to be instantiated by itself

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    @Id
+    @Column(name = "juror_number")
+    @StringSequenceGenerator(format = "%09d",
+            sequenceGenerator = @SequenceGenerator(id = "juror_number", start = 1)
+    )
+    private String jurorNumber;
+
+    @Column(name = "date_received")
+    private LocalDate dateReceived;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: Needs date generator

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    @Column(name = "first_name")
+    @FirstNameGenerator
+    private String firstName;
+
+    @Column(name = "last_name")
+    @LastNameGenerator
+    private String lastName;
+
+    @Column(name = "processing_status")
+    private String processingStatus;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: Needs default value

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Juror email address.
+     */
+    @Column(name = "email")
+    @EmailGenerator
+    private String email;
+
+    @Column(name = "residency")
+    private Boolean residency;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: All columns need default can be done using @FixedValueGenerator("your default")

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/Address.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+@MappedSuperclass
+@Getter
+@Setter
+@EqualsAndHashCode
+public abstract class Address implements Serializable {
+
+    @Column(name = "address_line_1")
+    @NotBlank
+    @RegexGenerator(regex = "[0-9]{0,9} [A-Z]{5,15} (Road|Lane|Gate|Close|Avenue|Street|Way|Drive|Gardens|Crescent|"
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/Address.java]: Regex can be moved to Constants class

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/Address.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            + "Park|Parklands|Parkway|Parade|Paradise|Parc|Parks|Part|Passage|Path|Pathway|Pike|Pines|Place|"
+            + "Plain|Plains|Plaza|Point|Points|Port|Ports|Prairie|Quadrant|Quay|Quays|Ramble|Ramp|Ranch|Rapids|"
+            + "Reach|Reserve|Rest|Retreat|Ridge|Ridges|Rise|River|Rivers|Road|Roads|Route|Row|Rue|Run|Shoal|"
+            + "Shoals|Shore|Shores|Skyline|Slope|Slopes|Spur|Square|Squares|Stairs|Stead|Strand|Stream|Street|"
+            + "Streets|Summit|Terrace|Throughway|Trace|Track|Trafficway|Trail|Trailer|Tunnel|Turnpike|Underpass|"
+            + "Union|Unions|Valley|Valleys|Viaduct|View|Views|Village|Villages|Ville|Vista|Walk|Wall|Way|Ways|")
+    private String addressLine1;
+
+    @Column(name = "address_line_2")
+    @RegexGenerator(regex = "(Apartment|Suite|Room|Floor|Box) Number [0-9]{1,3}")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/Address.java]: Regex can be moved to Constants class

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+@AllArgsConstructor
+@Getter
+@Setter
+//@DigitalThirdPartyOtherReason
+@Table(name = "juror_response", schema = "juror_mod")
+@EqualsAndHashCode(callSuper = true)
+public class DigitalResponse extends AbstractJurorResponse {
+
+    @Column(name = "residency")
+    @Builder.Default
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Needs default using @FixedValueGenerator.
(Same for all fields with no generator)

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/PaperResponse.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+@AllArgsConstructor
+@Builder
+@Getter
+@Setter
+@Table(name = "juror_response", schema = "juror_mod")
+@EqualsAndHashCode(callSuper = true)
+public class PaperResponse extends AbstractJurorResponse {
+
+    @Column(name = "mental_health_capacity")
+    private Boolean mentalHealthCapacity;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/PaperResponse.java:26]: Needs defaults

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReasonableAdjustments.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import jakarta.persistence.Table;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@NoArgsConstructor
+@AllArgsConstructor
+@Table(name = "t_reasonable_adjustments", schema = "juror_mod")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReasonableAdjustments.java]: Any constant table (annotation with t_) is not needed

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReplyType.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import jakarta.persistence.Table;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+import java.io.Serializable;
+
+@Entity
+@Table(name = "t_reply_type", schema = "juror_mod")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReplyType.java]: Class not needed as we can use Id's instead

File: src/main/java/uk/gov/hmcts/juror/support/sql/repository/JurorDigitalResponseRepository.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import uk.gov.hmcts.juror.support.sql.entity.jurorresponse.DigitalResponse;
+import uk.gov.hmcts.juror.support.sql.entity.jurorresponse.ReplyType;
+
+@Repository
+public interface JurorDigitalResponseRepository extends JurorResponseRepository<DigitalResponse> {
+    DigitalResponse findByJurorNumberAndReplyType(String jurorNumber, ReplyType replyType);
+
+    default DigitalResponse findByJurorNumber(String jurorNumber) {
+        return findByJurorNumberAndReplyType(jurorNumber, new ReplyType("Digital", null));
+    }
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/repository/JurorDigitalResponseRepository.java]: We only use repositories to save as such no methods are needed

File: src/main/java/uk/gov/hmcts/juror/support/sql/repository/JurorPaperResponseRepository.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+import uk.gov.hmcts.juror.support.sql.entity.jurorresponse.PaperResponse;
+import uk.gov.hmcts.juror.support.sql.entity.jurorresponse.ReplyType;
+
+@Repository
+public interface JurorPaperResponseRepository extends JurorResponseRepository<PaperResponse> {
+    PaperResponse findByJurorNumberAndReplyType(String jurorNumber, ReplyType replyType);
+
+    default PaperResponse findByJurorNumber(String jurorNumber) {
+        return findByJurorNumberAndReplyType(jurorNumber, new ReplyType("Paper", null));
+    }
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/repository/JurorPaperResponseRepository.java]: Methods not needed

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Column(name = "deferral_date")
+    @LocalDateGenerator(
+            minInclusive = @DateFilter(mode = DateFilter.Mode.PLUS, value = 4, unit = ChronoUnit.MONTHS),
+            maxExclusive = @DateFilter(mode = DateFilter.Mode.PLUS, value = 2, unit = ChronoUnit.WEEKS)
+    )
+    private String deferralDate;
+
+    @Column(name = "excusal_reason")
+    @FixedValueGenerator("d")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    @FixedValueGenerator("false")
+    private String bailDetails;
+
+    @Column(name = "convictions_details")
+    @FixedValueGenerator("false")
+    private String convictionsDetails;
+
+    @Column(name = "deferral_reason")
+    @FixedValueGenerator("false")
+    private String deferralReason;
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java:69]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    @Column(name = "convictions")
+    @FixedValueGenerator("false")
+    private Boolean convictions = Boolean.FALSE;
+
+    @Column(name = "mental_health_act")
+    @FixedValueGenerator("false")
+    private Boolean mentalHealthAct = Boolean.FALSE;
+
+    @Column(name = "residency_detail")
+    @FixedValueGenerator("true")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    private String residencyDetail;
+
+    @Column(name = "mental_health_act_details")
+    @FixedValueGenerator("false")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    private String mentalHealthActDetails;
+
+    @Column(name = "bail_details")
+    @FixedValueGenerator("false")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    private String bailDetails;
+
+    @Column(name = "convictions_details")
+    @FixedValueGenerator("false")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Column(name = "deferral_reason")
+    @FixedValueGenerator("false")
+    private String deferralReason;
+
+    @Column(name = "deferral_date")
+    @LocalDateGenerator(
+            minInclusive = @DateFilter(mode = DateFilter.Mode.PLUS, value = 4, unit = ChronoUnit.MONTHS),
+            maxExclusive = @DateFilter(mode = DateFilter.Mode.PLUS, value = 2, unit = ChronoUnit.WEEKS)
+    )
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    @FixedValueGenerator("d")
+    private String excusalReason;
+
+    @Version
+    @Column(name = "version")
+    @FixedValueGenerator("1")
+    private Integer version;
+
+    @Column(name = "thirdparty_fname")
+    @FirstNameGenerator
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+    private String thirdPartyFName;
+
+    @Column(name = "thirdparty_lname")
+    @LastNameGenerator
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    @RegexGenerator(regex = Constants.PHONE_REGEX)
+    private String otherPhone;
+
+    @Column(name = "email_address")
+    @EmailGenerator
+    private String emailAddress;
+
+    @Column(name = "thirdparty_other_reason")
+//    @Pattern(regexp = NO_PIPES_REGEX)
+    @FixedValueGenerator("ill")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to null

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+    private String thirdPartyOtherReason;
+
+    @Column(name = "juror_phone_details")
+    @Builder.Default
+    @RegexGenerator(regex = Constants.PHONE_REGEX)
+    private Boolean jurorPhoneDetails = Boolean.TRUE;
+
+    @Column(name = "juror_email_details")
+    @Builder.Default
+    @EmailGenerator
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/DigitalResponse.java]: Should default to true

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReasonableAdjustments.java [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+@Getter
+@Setter
+public class ReasonableAdjustments {
+    @Column(name = "code")
+    @Id
+    private String code;
+
+    @Column(name = "description")
+    private String description;
+}
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/support/sql/entity/jurorresponse/ReasonableAdjustments.java]: Not needed can use id

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/juror/support/sql/entity/AbstractJurorResponse.java]: Should be: 

  @Column(name = "processing_status")
    @Enumerated(EnumType.STRING)
    @FixedValueGenerator("ProcessingStatus.TODO")
    private ProcessingStatus processingStatus = ProcessingStatus.TODO;

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

**Review by Matt-Smith-CGI — COMMENTED** [review]
reviewed

**Review by Ben-Edwards-cgi — APPROVED** [review]

**Pull request merged**